xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1279,10 +1268,210 @@ public void run() {
+    }
+
+    String typeUrlV2() {
+      switch (this) {
+        case LDS:
+          return ADS_TYPE_URL_LDS_V2;
+        case RDS:
+          return ADS_TYPE_URL_RDS_V2;
+        case CDS:
+          return ADS_TYPE_URL_CDS_V2;
+        case EDS:
+          return ADS_TYPE_URL_EDS_V2;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown or missing case in enum switch: " + this);
+      }
+    }
+
+    static ResourceType fromTypeUrl(String typeUrl) {
+      switch (typeUrl) {
+        case ADS_TYPE_URL_LDS:
+          // fall trough
+        case ADS_TYPE_URL_LDS_V2:
+          return LDS;
+        case ADS_TYPE_URL_RDS:
+          // fall through
+        case ADS_TYPE_URL_RDS_V2:
+          return RDS;
+        case ADS_TYPE_URL_CDS:
+          // fall through
+        case ADS_TYPE_URL_CDS_V2:
+          return CDS;
+        case ADS_TYPE_URL_EDS:
+          // fall through
+        case ADS_TYPE_URL_EDS_V2:
+          return EDS;
+        default:
+          return UNKNOWN;
+      }
+    }
+  }
 
-    private StreamObserver<DiscoveryRequest> requestWriter;
+  private static final class DiscoveryRequestData {
+    private final ResourceType resourceType;
+    private final List<String> resourceNames;
+    private final String versionInfo;
+    private final String responseNonce;
+    private final Node node;
+    @Nullable
+    private final com.google.rpc.Status errorDetail;
+
+    private DiscoveryRequestData(
+        ResourceType resourceType, List<String> resourceNames, String versionInfo,
+        String responseNonce, Node node, @Nullable com.google.rpc.Status errorDetail) {
+      this.resourceType = resourceType;
+      this.resourceNames = resourceNames;
+      this.versionInfo = versionInfo;
+      this.responseNonce = responseNonce;
+      this.node = node;
+      this.errorDetail = errorDetail;
+    }
+
+    DiscoveryRequest toEnvoyProto() {
+      DiscoveryRequest.Builder builder =
+          DiscoveryRequest.newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node.toEnvoyProtoNode())
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(resourceType.typeUrl())
+              .setResponseNonce(responseNonce);
+      if (errorDetail != null) {
+        builder.setErrorDetail(errorDetail);
+      }
+      return builder.build();
+    }
+
+    io.envoyproxy.envoy.api.v2.DiscoveryRequest toEnvoyProtoV2() {
+      io.envoyproxy.envoy.api.v2.DiscoveryRequest.Builder builder =
+          io.envoyproxy.envoy.api.v2.DiscoveryRequest.newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node.toEnvoyProtoNodeV2())
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(resourceType.typeUrlV2())
+              .setResponseNonce(responseNonce);
+      if (errorDetail != null) {
+        builder.setErrorDetail(errorDetail);
+      }
+      return builder.build();
+    }
+
+    static Builder newBuilder() {
```

Comment:
nit: Builder may not be necessary given the usage is private. But up to you.

Reply:
Done.